src/reducers/importText.ts

```diff
@@ -157,22 +160,55 @@ const importText = (state: State, { path, text, lastUpdated, preventSetCursor, r
 
   }
   else {
+
     const json = isRoam ? roamJsonToBlocks(JSON.parse(convertedText)) : convertHTMLtoJSON(convertedText)
-    const imported = importJSON(state, simplePath, json, { lastUpdated, skipRoot })
 
-    return reducerFlow([
+    const uuid = createId()
 
-      updateThoughts(imported),
+    // Note: Create a dummy thought and then import new thoughts into its context and then collpase it . Since collapse uses existingThoughtMove it merges imported thoughts.
+    const updatedState = newThought(state, {
+      at: simplePath,
+      insertNewSubthought: true,
+      value: uuid
+    })
 
-      // restore the selection to the last imported thought on the first level
-      !preventSetCursor
-        ? setCursor({
-          path: imported.lastImported,
-          offset: headValue(imported.lastImported).length
-        })
-        : null,
+    const destEmpty = destThought.value === '' && getAllChildren(state, pathToContext(simplePath)).length === 0
 
-    ])(state)
+    // New dummy thoughts for collapsing
+    const newDummyThought = getAllChildren(updatedState, pathToContext(simplePath)).find(child => child.value === uuid)
+
+    const newDestinationPath = unroot(newDummyThought ? simplePath.concat(newDummyThought) : simplePath)
+
+    const imported = importJSON(updatedState, newDestinationPath as SimplePath, json, { lastUpdated, skipRoot })
+
+    /** Set cursor to the last imported path. */
+    const setLastImportedCursor = (state: State) => {
+      const cursorContextHead = pathToContext(imported.lastImported).slice(0, newDestinationPath.length - (destEmpty ? 2 : 1))
+      const cursorContextTail = pathToContext(imported.lastImported).slice(newDestinationPath.length)
+      const newCursorContext = cursorContextHead.concat(cursorContextTail)
+      const newCursor = rankThoughtsFirstMatch(state, newCursorContext)
+
+      return setCursor(state, {
+        path: newCursor
+      })
+    }
+
+    return reducerFlow([
+      updateThoughts(imported),
+      // set cursor to the dummy thought
+      setCursor({
+        path: newDestinationPath,
+      }),
+      collapseContext({ preventCursorArchive: true }),
+      // if original destination has empty then collapse once more
+      ...destEmpty ? [setCursor({
+        path: parentOf(newDestinationPath)
+      }), collapseContext({ preventCursorArchive: true })] : [],
```

Comment:
I'm concerned about the use of `setCursor` when `preventSetCursor` is true. What about adding an `at` option to `collapseContext` that defaults to `state.cursor` but can be set to any `Path`? Then we can pass `parentOf(newDestinationPath)` directly. We use this pattern elsewhere so there is some precedent.